### PR TITLE
Docs for turning assembled haplotypes into variant alleles

### PR DESCRIPTION
**Pull request by davidbenjamin**
@takutosato This wraps up the documentation part of our quarterly goals.

### CODE REVIEW

File: docs/variants_from_haplotypes.tex [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Taking all the unique start positions and variant alleles from all haplotypes give an initial set of variants to genotype but we are not quite done.  If multiple haplotypes have different variant alleles at the same position we may nee to reconcile the representations.  For example, we may have a single deletion $AA \rightarrow A$ and a double deletion $AAA \rightarrow A$, which need to be merged as $AAA \rightarrow A, AA$.  Fortunately, this is essentially the canonical example and there are no edge cases to deal with.  That is, all we need to do is find the allele with the longest reference, and pad the other alleles with whatever of these extra reference bases they were missing.
```
Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: need* to

File: docs/variants_from_haplotypes.tex [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
...at is the maximum among the bottom row and rightmost column of $M$.  If this maximum is in the rightmost column that means all alternate haplotype bases are used and nothing special must be done.  If, however the maximum is on the bottom row it means that the more alternate bases remain.  In this case, we record a soft clip CIGAR element with length equal to the number of remaining bases at the end of the alignment\footnote{The code behaves differently for different values of the \code{OverhandStrategy} enum, but this strategy is hard-coded to \code{SOFTCLIP}, which results in the behavior we describe here.}.  Then, from this $(i,j)$ we iterate the following procedure until reaching $i = 0$ or $j = 0$ (recall that these correspond to immediately \textit{before} the start of the corresponding haplotypes): If $B_{ij} = 0$ add a match (M) CIGAR element to the left end of the alignment and move to $i,j \rightarrow i - 1, j - 1$.  If $B_{ij} = k > 0$ add a length-$k$ deletion CIGAR element to the left of the alignment and move to $i,j \rightarrow i - k, j$.  If $B_{ij} = -k < 0$ add a length-$k$ insertion CIGAR element to the left of the alignment and move to $i,j \rightarrow i - k, j$.
```
Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: for insertion, `$i, j \rightarrow i, j-k$`

Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: I liked that you added (M) to "add a match (M) CIGAR element..." Maybe you can do the same for deletion (D) CIGAR element, insertion (I) CIGAR element, and soft clip (S) CIGAR element?

File: docs/variants_from_haplotypes.tex [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+\item Record backtrack: The score for no gap, i.e. a match \textit{alignment} (as opposed to matching bases), is $M_{i-1,j-1}$ plus $w_{\rm match}$ if the $i$th reference base and $j$th alternate base agree or $w_{\rm mismatch}$ if they do not.  We now compare this score to the indel scores $D_j$ and $R$.  We set $B_{ij}$ to 0 if the match score is greatest, $S^d_j$ if the downward gap score is greatest, and $-S^r$ if the rightward gap score is greatest.  This convention essentially uses the sign an as enum in order to encode whether the optimal path has an insertion, deletion or match.
```
Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: *as an enum

File: docs/variants_from_haplotypes.tex [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+First we initialize the zeroth row and column as $M_{0,0}$ = 0, $M_{0,1} = M_{1,0} = w_{\rm open}$, $M_{0,2} = M_{2,0} = w_{\rm open} + w{\rm extend}$ etc.  The zeroth row and column correspond to one base before the reference and alternate starts and this initialization penalizes leading indels.  That is, this is a global alignment\footnote{The GATK assembly graph merges all alternate paths into the reference, hence alternate haplotypes start and end coincident with the reference and global alignment is appropriate.}.
+
+Then, for each row $1 \le i \le {\rm length(reference)}$ we loop over all columns $1 \le j \le {\rm length(alternate)}$ and do the following:
+
+\begin{itemize}
+\item Update deletion scores:  The score for opening a downward gap is $M_{i-1,j} + w_{\rm open}$.  The score for extending an existing deletion is $G_j + w{\rm extend}$, where $D_j$ at this point still pertains to the $previous$ row $i - 1$.  We set $D_j$ (modifying it in-place) to the greater of these values. If the gap-opening score is greater than the gap-extending score, we set $S^d_j = 1$, otherwise we increment $S^d_j$ by 1.
```
Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: existing deletion is *`$D_j + w_{\rm extend}$`

File: docs/variants_from_haplotypes.tex [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+\item Update insertion scores:  When we begin traversing row $i$ we initialize the current best score for alignments ending in a ``rightward'' gap (i.e. an insertion) as $R = -\infty$ and we initialize the length of the terminal gap in this best-scoring alignment as $S^r = 0$.  Note that these values are local to the inner loop over $j$ and thus do not need to be arrays\footnote{In the code, they \textit{are} arrays, but in the $i$th iteration of the loop over rows, only their $i$th elements are used.  Thus the elements are effectively scalars whose scope is the loop over $j$.}.  At each stage in the loop over $j$ the score for opening a rightward gap is $M_{i,j - 1} + w_{\rm open}$ and score for extending one is $R + w{\rm extend}$, where $R$ still pertains to the previous column $j - 1$.  We set $R$ (modifying it in-place)  to the greater of these values.  If the gap-opening score is greater than the gap-extending score, we set $S^r_j = 1$, otherwise we increment $S^r_j$ by 1.
```
Comment (takutosato) [inline docs/variants_from_haplotypes.tex]: This sounds to me like we don't allow multiple insertions or multiple deletions in a single read - is that right? Maybe that's worth a mention

File: docs/variants_from_haplotypes.tex [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+Despite its name, \HC does not actually call haplotypes.  Rather, it generates haplotypes as an intermediate step to discover variants at individual loci.  Here we describe how the GATK engine determines which alt alleles exist in locally assembled haplotypes.
+\end{abstract}
+
+\maketitle
+
+\section{Forward pass}
+The first step is to align each assembled haplotype to the reference haplotype using the Smith-Waterman algorithm.  Although the GATK's implementation is not complicated, it is also not a completely direct translation of the method into code.  As it incurs a non-trivial computational cost, we describe it in detail here.
+
+Our implementation has four score parameters, $w_{\rm match}$ and $w_{\rm mismatch}$ for equal and unequal reference and alternate bases, $w_{\rm open}$ for opening a gap (that is, starting an indel), and $w{\rm extend}$ for extending a gap.  Note the absence of a scoring matrix treating each possible type of match and mismatch differently.  Although the idea of a score seems like a heuristic, the algorithm is equivalent to finding the maximum likelihood path in a hidden Markov model in which the scores are log transition and emission probabilities.
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: done in all places

File: docs/variants_from_haplotypes.tex (hunk 4 above)
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: fixed

File: docs/variants_from_haplotypes.tex (hunk 5 above)
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: We allow multiple indels.  I think the confusion is about the definition of the M_ij matrix.  This is the best score of alignments that are in a match state *at i,j* but which could have indels earlier.  I added a footnote making this explicit.

File: docs/variants_from_haplotypes.tex (hunk 3 above)
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: fixed

File: docs/variants_from_haplotypes.tex (hunk 2 above)
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: nice catch, fixed

Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: done

File: docs/variants_from_haplotypes.tex (hunk 1 above)
Comment (davidbenjamin) [reply docs/variants_from_haplotypes.tex]: fixed

File: docs/variants_from_haplotypes.tex (hunk 5 above)
Comment (takutosato) [reply docs/variants_from_haplotypes.tex]: Got it, thanks!

**Pull request merged**